examples/vision/nerf.py

```diff
@@ -0,0 +1,676 @@
+    return (rgb, depth_map)
+
+
+"""
+# "Model.fit"
+
+The training step is gathered in a `keras.Model` class so that we can
+make use of the `model.fit` call.
+"""
+
+
+class NeRF(keras.Model):
+    def __init__(self, nerf_model):
+        super().__init__()
+        self.nerf_model = nerf_model
+
+    def compile(self, optimizer, loss):
+        super().compile()
+        self.optimizer = optimizer
+        self.loss = loss
+
+    def train_step(self, inputs):
+        # Get the images and the rays.
+        (images, rays) = inputs
+        (rays_flat, t_vals) = rays
+
+        with tf.GradientTape() as tape:
+            # Get the predictions from the model.
+            rgb, _ = render_rgb_depth(
+                model=self.nerf_model, rays_flat=rays_flat, t_vals=t_vals, rand=True
+            )
+            loss = self.loss(images, rgb)
+
+        # Get the trainable variables.
+        trainable_variables = self.nerf_model.trainable_variables
+
+        # Get the gradeints of the trainiable variables with respect to the loss.
+        gradients = tape.gradient(loss, trainable_variables)
+
+        # Apply the grads and optimize the model.
+        self.optimizer.apply_gradients(zip(gradients, trainable_variables))
+
+        # Get the PSNR of the reconstructed images and the source images.
+        psnr = tf.image.psnr(images, rgb, max_val=1.0)
+
+        # return the loss and the psnr value
+        return {"loss": loss, "psnr": psnr}
+
+
+test_imgs, test_rays = next(iter(train_ds))
+test_rays_flat, test_t_vals = test_rays
+
+loss_list = []
+
+
+class CustomCallback(keras.callbacks.Callback):
+    def on_epoch_end(self, epoch, logs=None):
+        loss = logs["loss"]
+        loss_list.append(loss)
+        test_recons_images, depth_maps = render_rgb_depth(
+            model=self.model.nerf_model,
+            rays_flat=test_rays_flat,
+            t_vals=test_t_vals,
+            rand=True,
+            train=False,
+        )
+
+        # Plot the rgb, depth and the loss plot.
+        fig, ax = plt.subplots(nrows=1, ncols=3, figsize=(20, 5))
+        ax[0].imshow(keras.preprocessing.image.array_to_img(test_recons_images[0]))
+        ax[0].set_title(f"Predicted Image: {epoch:03d}")
+
+        ax[1].imshow(keras.preprocessing.image.array_to_img(depth_maps[0, ..., None]))
+        ax[1].set_title(f"Depth Map: {epoch:03d}")
+
+        ax[2].plot(loss_list)
+        ax[2].set_xticks(np.arange(0, EPOCHS + 1, 5.0))
+        ax[2].set_title(f"Loss Plot: {epoch:03d}")
+
+        fig.savefig(f"images/{epoch:03d}.png")
+        plt.close()
+
+
+num_pos = H * W * NUM_SAMPLES
+nerf_model = get_nerf_model(num_layers=8, num_pos=num_pos)
+
+model = NeRF(nerf_model)
+model.compile(optimizer=keras.optimizers.Adam(), loss=keras.losses.MeanSquaredError())
+
+# Create a directory to save the images during training.
+if not os.path.exists("images"):
+    os.makedirs("images")
+
+model.fit(
```

Comment:
You should monitor generalization during `fit()` by using a separate `validation_data` dataset.